From 40c914a2c07126233eafc1c388acb03887589871 Mon Sep 17 00:00:00 2001
From: Maksim Sisov <msisov@igalia.com>
Date: Mon, 6 Aug 2018 11:02:24 +0300
Subject: [PATCH] Fix kiosk and start-maximized modes in AGL.

For some reason, weston gets stuck when chromium start with kiosk
or start maximized mode until the surface is activated.

To avoid this, set the maximize and fullscreen to be pending until
surface is activated.
---
 ui/ozone/platform/wayland/wayland_window.cc | 20 ++++++++++++++++++++
 ui/ozone/platform/wayland/wayland_window.h  |  4 ++++
 2 files changed, 24 insertions(+)

diff --git a/ui/ozone/platform/wayland/wayland_window.cc b/ui/ozone/platform/wayland/wayland_window.cc
index ca376f810b41..dfbacb84a34a 100644
--- a/ui/ozone/platform/wayland/wayland_window.cc
+++ b/ui/ozone/platform/wayland/wayland_window.cc
@@ -277,6 +277,11 @@ bool WaylandWindow::HasCapture() const {
 void WaylandWindow::ToggleFullscreen() {
   DCHECK(xdg_surface_);
 
+  if (!is_active_) {
+    pending_fullscreen_ = true;
+    return;
+  }
+
   // TODO(msisov, tonikitoo): add multiscreen support. As the documentation says
   // if xdg_surface_set_fullscreen() is not provided with wl_output, it's up to
   // the compositor to choose which display will be used to map this surface.
@@ -308,6 +313,11 @@ void WaylandWindow::ToggleFullscreen() {
 void WaylandWindow::Maximize() {
   DCHECK(xdg_surface_);
 
+  if (!is_active_) {
+    pending_maximized_ = true;
+    return;
+  }
+
   if (IsFullscreen())
     ToggleFullscreen();
 
@@ -506,6 +516,16 @@ void WaylandWindow::HandleSurfaceConfigure(int32_t width,
 
   if (did_active_change)
     delegate_->OnActivationChanged(is_active_);
+
+  if (!was_active_ && is_active_) {
+    if (pending_fullscreen_) {
+      ToggleFullscreen();
+    } else if (pending_maximized_) {
+      Maximize();
+    }
+    pending_fullscreen_ = false;
+    pending_maximized_ = false;
+  }
 }
 
 void WaylandWindow::OnCloseRequest() {
diff --git a/ui/ozone/platform/wayland/wayland_window.h b/ui/ozone/platform/wayland/wayland_window.h
index b8381f5449b1..4abd35844454 100644
--- a/ui/ozone/platform/wayland/wayland_window.h
+++ b/ui/ozone/platform/wayland/wayland_window.h
@@ -161,10 +161,14 @@ class WaylandWindow : public PlatformWindow, public PlatformEventDispatcher {
   ui::PlatformWindowState state_;
 
   bool is_active_ = false;
+  bool was_active_ = false;
   bool is_minimizing_ = false;
 
   bool is_tooltip_ = false;
 
+  bool pending_fullscreen_ = false;
+  bool pending_maximized_ = false;
+
   DISALLOW_COPY_AND_ASSIGN(WaylandWindow);
 };
 
-- 
2.11.0

